hibernate-core/src/main/java/org/hibernate/type/CollectionType.java

```diff
@@ -712,7 +712,12 @@ public Object replace(
 			final Object owner,
 			final Map copyCache) throws HibernateException {
 		if ( original == null ) {
-			return null;
+			if(target instanceof PersistentCollection) {
+				return target;
+			}
+			else {
+				return null;
+			}
```

Comment:
Copying the conversation from [the original PR](https://github.com/hibernate/hibernate-orm/pull/4339#discussion_r742621019).

From @yrodiere:

> So if someone erased a collection in a entity by setting it to `null`, and then tries to merge that entity, the collection will all of a sudden no longer be erased.
> 
> This looks like you're changing the meaning of a `null` collection from "empty" to "undefined". While I could get behind that in a new ORM implementation, doing this in Hibernate ORM will definitely break many existing application.

From @busitech:

> Your first statement in this comment is incorrect, not only because this PR is passing almost all test cases in Hibernate Core, but also because cascade processing has already occurred by this time.  
> 
> In the above method, if `target` is a `PersistentCollection` and the original value was `null`, we are currently processing a lazy loaded collection property that Hibernate Search may need to traverse.  We need to assign the uninitialized proxy value to the object.  The purpose of this code is to not overlay an uninitialized proxy collection with the originally null value, so Hibernate Search can traverse the relationship, and possibly go deeper into the hierarchical tree by traversing even beyond properties of the present entity instance.

From @busitech:

> It is an incorrect assumption that `null` always means empty, even in ORM today, and this ambiguity is the root cause of our problems.  ORM needs to change this to resolve HHH-14839 with a long term solution.  As discussed previously in the forum, this PR is a workaround pending a fundamental architecture change to ORM.
> 
> ORM is inconsistent in that it _NEVER_ returns a `null` value for a lazy loaded collection property after loading an entity, but _ALWAYS_ returns `null` collection values passed in to `em.persist` and `em.merge`, even when they should have data.  
>
> **The main effect of this PR is to standardize results returned from `em.persist` and `em.merge` so that they are more consistent with those of `em.find`.**
> 
> ORM cannot carry forward bad design forever.  At some point it must be realized that a breaking change is absolutely necessary to purge bad design.  Long term, if a `null` collection will not mean uninitialized, then there needs to be a placeholder for collections that actually means uninitialized in all cases.
> 
> It is our opinion that the meaning of `null` must be changed from a concept of "empty" to "undefined" **_when assigned to a collection property_**, and that only a collection object with length zero should be seen as an empty collection.
> 
> Note also that at the present time, the code in this PR does not affect associations which cascade in any way, to limit the complexity initially.  It might be beneficial to Hibernate Search to have all cascading collections transformed into an uninitialized proxy as well, but we hardly ever use them in our application, so we did not address this yet.
> 
> If we _were_ to apply this logic to collections that use cascade features, applications would simply see a more valid way of representing an empty collection returned instead of `null`.  Applications would see an actual collection object with length of zero, instead of what was previously `null` and ambiguous.
>
> If such an application were depending on seeing null instead of a collection with zero length, it would have to change slightly, but as I already stated, ORM _already_ provides applications with such a value for ALL lazy collection properties during `em.find`, so they are already dealing with them there...

Reply:
I heard you say you could get behind it for a new ORM implementation, which I took to mean a new alternative product other than Hibernate ORM.  I thought you meant it would be an eternal dead end trying to get Hibernate ORM improved.  If you meant getting behind it for ORM 6 then that would be terrific.

Let's pursue option 1 on your end, and on this end we are going to run our build of 5 in production, immediately.  We will be able to benefit from our own build while 6 is updated as well.

Reply:
I am working on porting our PR to Hibernate 6.